File: python/src/skills_utils/sync.py
from __future__ import annotations

import shutil
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from skills_utils.agents import resolve_skills_dir
from skills_utils.config_loader import set_skill_entry, utc_now_iso
from skills_utils.hash import dir_hash
from skills_utils.progress import SyncProgress
from skills_utils.repos import RemoteSkill, RepoConfig, download_skill_files, repo_cache_dir
from skills_utils.github import GitHubClient
from skills_utils.scanner import flat_symlink_path, install_key, nested_install_path


@dataclass
class SyncPlan:
    install: list[str]
    remove: list[str]
    update: list[str]


@dataclass
class SyncResult:
    installed: list[str] = None  # type: ignore[assignment]
    removed: list[str] = None  # type: ignore[assignment]
    updated: list[str] = None  # type: ignore[assignment]

    def __post_init__(self) -> None:
        self.installed = self.installed or []
        self.removed = self.removed or []
        self.updated = self.updated or []


def apply_sync(
    client: GitHubClient,
    repo: RepoConfig,
    agent_id: str,
    agent: dict[str, Any],
    manifest: dict[str, Any],
    remote_inventory: dict[str, RemoteSkill],
    commit_sha: str,
    selected: set[str],
    to_update: set[str],
    flat_symlinks: bool,
    skills_dir: Path | None = None,
    progress: SyncProgress | None = None,
) -> SyncResult:
    target_dir = skills_dir or resolve_skills_dir(agent)
    target_dir.mkdir(parents=True, exist_ok=True)
    result = SyncResult()
    existing = {
        name
        for name, remote in remote_inventory.items()
        if nested_install_path(target_dir, repo.slug, name).exists()
    }

    to_remove = sorted(existing - selected)
    to_apply: list[tuple[str, RemoteSkill]] = []
    for name in sorted(selected):
        remote = remote_inventory[name]
        install_path = nested_install_path(target_dir, repo.slug, name)
        was_installed = install_path.exists() and (install_path / "SKILL.md").exists()
        key = install_key(repo.slug, name)
        entry = manifest.get("agents", {}).get(agent_id, {}).get(key)
        needs_copy = (
            not was_installed
            or name in to_update
            or not entry
            or entry.get("remote_fingerprint") != remote.fingerprint
        )
        if needs_copy:
            to_apply.append((name, remote))
        elif flat_symlinks:
            _ensure_flat_symlink(target_dir, repo.slug, name)

    if progress:
        progress.begin(len(to_remove) + len(to_apply))

    for index, name in enumerate(to_remove, start=1):
        if progress:
            progress.removing(name, index, len(to_remove))
        _remove_skill(target_dir, repo.slug, name, agent_id, manifest)
        result.removed.append(name)

    for name, remote in to_apply:
        install_path = nested_install_path(target_dir, repo.slug, name)
        was_installed = install_path.exists() and (install_path / "SKILL.md").exists()

        cache_skill_dir = repo_cache_dir(repo) / repo.skills_path / name
        download_skill_files(client, repo, remote, cache_skill_dir, progress=progress)
        if progress:
            progress.installing(name)
        if install_path.exists():
            shutil.rmtree(install_path)
        shutil.copytree(cache_skill_dir, install_path)

        if not was_installed:
            result.installed.append(name)
        else:
            result.updated.append(name)

        set_skill_entry(
            manifest,
            agent_id,
            install_key(repo.slug, name),
            {
                "repo_slug": repo.slug,
                "remote_path": remote.remote_path,
                "commit_sha": commit_sha,
                "remote_fingerprint": remote.fingerprint,
                "installed_hash": dir_hash(install_path),
                "installed_at": utc_now_iso(),
            },
        )
        if flat_symlinks:
            if progress:
                progress.linking(name)
            _ensure_flat_symlink(target_dir, repo.slug, name)

    return result


def _remove_skill(
    skills_dir: Path, repo_slug: str, skill_name: str, agent_id: str, manifest: dict[str, Any]
) -> None:
    path = nested_install_path(skills_dir, repo_slug, skill_name)
    if path.exists():
        shutil.rmtree(path)
    link = flat_symlink_path(skills_dir, skill_name)
    if link.is_symlink():
        link.unlink()
    from skills_utils.config_loader import remove_skill_entry

    remove_skill_entry(manifest, agent_id, install_key(repo_slug, skill_name))


def _ensure_flat_symlink(skills_dir: Path, repo_slug: str, skill_name: str) -> None:
    link = flat_symlink_path(skills_dir, skill_name)
    target = Path(repo_slug) / skill_name
    if link.exists() or link.is_symlink():
        if link.is_symlink() and link.resolve() == (skills_dir / repo_slug / skill_name).resolve():
            return
        if link.exists():
            return
        if link.is_symlink():
            link.unlink()
    if link.exists():
        return
    link.symlink_to(target)

File: node/src/sync.ts
import { access, cp, mkdir, rm, symlink } from "node:fs/promises";
import path from "node:path";
import { utcNowIso } from "./config.js";
import { dirHash } from "./hash.js";
import type { GitHubClient } from "./github.js";
import type { SyncProgressReporter } from "./progress.js";
import { downloadSkillFiles, repoCacheDir, type RemoteSkill, type RepoConfig } from "./repos.js";
import { installKey, nestedInstallPath } from "./scanner.js";

export interface SyncResult {
  installed: string[];
  removed: string[];
  updated: string[];
}

export async function applySync(
  client: GitHubClient,
  repo: RepoConfig,
  agentId: string,
  manifest: any,
  remoteInventory: Record<string, RemoteSkill>,
  commitSha: string,
  selected: Set<string>,
  toUpdate: Set<string>,
  flatSymlinks: boolean,
  skillsDir: string,
  progress?: SyncProgressReporter,
): Promise<SyncResult> {
  const result: SyncResult = { installed: [], removed: [], updated: [] };
  await mkdir(skillsDir, { recursive: true });

  const existing: string[] = [];
  for (const name of Object.keys(remoteInventory)) {
    try {
      await access(path.join(nestedInstallPath(skillsDir, repo.slug, name), "SKILL.md"));
      existing.push(name);
    } catch {
      // not installed
    }
  }

  const toRemove = existing.filter((n) => !selected.has(n)).sort();
  const toApply: Array<[string, RemoteSkill]> = [];

  for (const name of [...selected].sort()) {
    const remote = remoteInventory[name];
    const installPath = nestedInstallPath(skillsDir, repo.slug, name);
    let wasInstalled = false;
    try {
      await access(path.join(installPath, "SKILL.md"));
      wasInstalled = true;
    } catch {
      wasInstalled = false;
    }
    const key = installKey(repo.slug, name);
    const entry = manifest.agents?.[agentId]?.[key];
    const needsCopy =
      !wasInstalled || toUpdate.has(name) || !entry || entry.remote_fingerprint !== remote.fingerprint;
    if (needsCopy) {
      toApply.push([name, remote]);
    } else if (flatSymlinks) {
      await ensureFlatSymlink(skillsDir, repo.slug, name);
    }
  }

  progress?.begin(toRemove.length + toApply.length);

  for (let index = 0; index < toRemove.length; index += 1) {
    const name = toRemove[index];
    progress?.removing(name, index + 1, toRemove.length);
    await removeSkill(skillsDir, repo.slug, name, agentId, manifest);
    result.removed.push(name);
  }

  for (const [name, remote] of toApply) {
    const installPath = nestedInstallPath(skillsDir, repo.slug, name);
    let wasInstalled = false;
    try {
      await access(path.join(installPath, "SKILL.md"));
      wasInstalled = true;
    } catch {
      wasInstalled = false;
    }

    const cacheSkillDir = path.join(repoCacheDir(repo), repo.skillsPath, name);
    await downloadSkillFiles(client, repo, remote, cacheSkillDir, progress);
    progress?.installing(name);
    await rm(installPath, { recursive: true, force: true });
    await cp(cacheSkillDir, installPath, { recursive: true });

    if (!wasInstalled) result.installed.push(name);
    else result.updated.push(name);

    manifest.agents ??= {};
    manifest.agents[agentId] ??= {};
    manifest.agents[agentId][installKey(repo.slug, name)] = {
      repo_slug: repo.slug,
      remote_path: remote.remotePath,
      commit_sha: commitSha,
      remote_fingerprint: remote.fingerprint,
      installed_hash: await dirHash(installPath),
      installed_at: utcNowIso(),
    };
    if (flatSymlinks) {
      progress?.linking(name);
      await ensureFlatSymlink(skillsDir, repo.slug, name);
    }
  }

  return result;
}

async function removeSkill(
  skillsDir: string,
  repoSlug: string,
  skillName: string,
  agentId: string,
  manifest: any,
): Promise<void> {
  const installPath = nestedInstallPath(skillsDir, repoSlug, skillName);
  await rm(installPath, { recursive: true, force: true });
  const link = path.join(skillsDir, skillName);
  await rm(link, { force: true });
  const key = installKey(repoSlug, skillName);
  delete manifest.agents?.[agentId]?.[key];
}

async function ensureFlatSymlink(
  skillsDir: string,
  repoSlug: string,
  skillName: string,
): Promise<void> {
  const link = path.join(skillsDir, skillName);
  const target = path.join(repoSlug, skillName);
  try {
    const { lstat } = await import("node:fs/promises");
    const info = await lstat(link);
    if (info.isSymbolicLink()) return;
    if (info.isDirectory() || info.isFile()) return;
  } catch {
    // create symlink
  }
  await rm(link, { force: true });
  await symlink(target, link);
}
